### ISSUE TITLE
Support stargazer

### ISSUE THREAD

**Issue by MaxGhenis** [bashtage/linearmodels#270]
The [`stargazer` Python package](https://github.com/mwburke/stargazer/) implements the [R package](https://cran.r-project.org/web/packages/stargazer/index.html) of the same name, which "Produces LaTeX code, HTML/CSS code and ASCII text for well-formatted tables that hold regression analysis results from several models side-by-side, as well as summary statistics."

Currently, it works with `statsmodels`, but not with `linearmodels` (https://github.com/mwburke/stargazer/issues/26).

In that issue, @toobaz [suggested](https://github.com/mwburke/stargazer/issues/26#issuecomment-626243710):
>I suspect it would be easy for linearmodels devs (or even an independent project) to build a compatibility layer which puts model internals where statsmodels expects them. I did something similar in https://github.com/toobaz/rmodel , which allows R models to be plugged into stargazer tables.

**Comment by bashtage** [comment]
I think this would probably need to go into stargazer or standalone package. Ideally, stargazer would have a well-defined API that a model shim could target to simplify things.

**Comment by toobaz** [comment]
> Ideally, stargazer would have a well-defined API

I agree, ideally. But in practice, stargazer is a young and small project, whereas statsmodels is a standard library in the field.

I have in mind something like (in this order):
1. interested projects provide their own compatibility interface to (an extension of) statsmodels objects' interface, so that stargazer works
2. statsmodels documents and possibly rearranges such interface
3. the interface possibly becomes a standard, not necessarily tied to statsmodels (and certainly not to stargazer - e.g. it can make sense also for sharing/transmitting/integrating results, not just for creating tables)

Alternatively: an independent project creates an interface and a translation for statsmodels. But that's a non-trivial task (e.g. stargazer currently supports very little of statsmodels' models). In any case, defining the API would be out of the scope of stargazer.

(I'm not speaking as, or on behalf of, the stargazer maintainer, just as a recurrent contributor)

I'll elaborate in https://github.com/mwburke/stargazer/issues/26 on the technical aspects.

**Comment by bashtage** [comment]
I think linermodels is unlikely to take stargazer as a dependency, which is what would be needed for this to go into linearmodels. It is also not feasible to change the linearmodels API. 

IMO there are a few sensible solutions:

1. stargazer could provide "adapters" for models from other packages(linearmodels, pysal, scikit-learn, and probably lots more), or other statsmodels models that don't fit the basic structure.  These adapters will take whatever if returned in the result of fitting the model and then repackage it to have a standard interface.
2. If the maintainer of stargazer doesn't want to do this, then it could go into another package (like RModel has done). For example, stargazer-adapters.
3. I suspect that it is pretty easy to write a simple shim object to wrap any specific class. ssuming stargazer duck types, then any object with properties like `params` and `bse` could be consumed. Although the lack of a well-defined API for stargazer makes this less fun.

I'm not suggesting that stargazer defines statsmodels API. Stargazer would define the API for model results that it would happily consume. This may look a lot like statsmodel OLSREsults, but could also differ. This is the most user-friendly approach since it lets any "target" stargazer for table production with a model estimated in anything simply by providing the object. In an ideal world, stargzaer would provide an `ABC` that would be wrapped, e.g.

```
from abc import ABCmeta, abstract_method

class StargazerResults(metaclass=ABCmeta):
   """ABC that projects can wrap to provide an object consumale by stargazer"""

    @property
    @abstract_method
    def params(self):
        """The estimated parameters, 1d"""

    @property
    @abstract_method
    def bse(self):
        """The parameter standard errors, 1d"""
...
```

**Comment by toobaz** [comment]
> I think linermodels is unlikely to take stargazer as a dependency, which is what would be needed for this to go into linearmodels

"this" what? I never suggested that linearmodels changes anything very deep in its code. I just suggested it (and any other project with similar needs) should either 1) provide a statsmodels-look-alike translation that stargazer can consume, or 2) make a list of what a statsmodels-look-alike translation lacks in order to do the task, and see if we can extend it.

... but apparently using statsmodels as a reference sounds humiliating, so forget it.

> 2\. If the maintainer of stargazer doesn't want to do this, then it could go into another package (like RModel has done). For example, stargazer-adapters.

As a stargazer contributor, I would object to this too. It still means that it is the responsability of the stargazer project, in a broader sense to, for instance, keep up with code changes in the various libraries. Moreover, as I said already, sharing model results among libraries goes beyond just creating tables - which is all stargazer does. So let's think about something independent from stargazer, for which stargazer can work as the reference test bench. Hint: the name "pymodref" is apparently free.

> stargazer would provide an `ABC` that would be wrapped

What is the advantage over just documenting what a library should expose in order to be compatible? Notice that virtually everything except params and dependent variables is (or should be) optional, so the API in the end should just, for each possible result that a model can produce,
- give it a name
- describe its possible type (including e.g. dimensions of arrays, and which dimensions of which arrays correspond)

... then each library could just expose a same method, e.g. ``_modref_get(feature_name)`` that stargazer (or anything else) would use.

I would be happy to support something like this in RModel, for instance.

If you think the approach makes sense, I can try to produce a new repo with a first draft.

**Comment by bashtage** [comment]
> "this" what? I never suggested that linearmodels changes anything very deep in its code. I just suggested it (and any other project with similar needs) should either 1) provide a statsmodels-look-alike translation that stargazer can consume, or 2) make a list of what a statsmodels-look-alike translation lacks in order to do the task, and see if we can extend it.

"this" is an adapter from a linearmodels result to something that stargazer would accept.  If something goes into linearmodels then linearmodels (or its maintainer) is responsible for making sure that it maintains compatibility with stargazer.  I don't think this is reasonable for such a small library like stargazer.  It makes more sense for stargazer to own the interface, or for it to be independent. This is similar to how pandas provides adapters (`pandas.io`) for all sorts of foreign data format that import into `DataFrame`s.

I looked at stargazer and it uses a `type(result) is RegressionResultsWrapper`. This makes it impossible to extend in a sensible way. 

> As a stargazer contributor, I would object to this too. It still means that it is the responsibility of the stargazer project, in a broader sense to, for instance, keep up with code changes in the various libraries. Moreover, as I said already, sharing model results among libraries goes beyond just creating tables - which is all stargazer does. So let's think about something independent from stargazer, for which stargazer can work as the reference test bench. Hint: the name "pymodref" is apparently free.

It makes more sense for statgazer to define what it wants.  Right not only wanting `RegressionResultsWrapper` makes it completely inflexible, as evidence by the hoops `RModels` had to jump through. 

> What is the advantage over just documenting what a library should expose in order to be compatible?

An ABC is a contract.  Override all of the abstract classes in the ABC with a user class and the consumer API promises that it will work. It also make version checks super simple, since all you need is `isinstance(results, StargazerResults`) since any child is required to subclass. It is much easier to have an ABC than to duck type. 

It is also OK for a child of an ABC to not return anything.  For example, the definition in the ABC could have something like

```
   @property
   @abstract_method
    def rsquare(self) -> Optional[float]:
        """The model's measure of fit, or None if not supported"""
```

I still think something like 

```
from statgazer.adapters.linearmodels import IVResult
# or statgazer.adapters.scikit_learn, or pysal...
# or from stargazer_adapters.MODULE import ModelType
from linearmodels.iv import IV2SLS

sg_res = IVResults(IV2SLS(y,x1,x2,z).fit())
```

provides the most straightforward way to users who want to use stargazer to extend it since it doesn't require tieing the upstream package to stargazer.

**Comment by toobaz** [comment]
> It makes more sense for stargazer to own the interface, or for it to be independent.

The latter.

Comments on my (last) proposal are welcome. It doesn't envisage ducktyping, ``RegressionResultsWrapper``, tying to stargazer or any other dangers you mentioned in your reply.

In any case...

> `from statgazer.adapters.linearmodels import IVResult`

I'm not following you here. This is different from your proposal of abstract class above, right?

**Comment by bashtage** [comment]
> The latter.
> 
> Comments on my (last) proposal are welcome. It doesn't envisage ducktyping, `RegressionResultsWrapper`, tying to stargazer or any other dangers you mentioned in your reply.

There will need to be _some_ changes in stargazer to make it a generally useful library, rather than a presentation layer for OLS and similar results. The most important would be to remove the explicit dependency on RegressionResultsWrapper. It would be pretty easy to 

```
from stargazer.adapters.abstract implement StargazerResults
from stargazer.adapters.statsmodels import RegressionResults

<in the main class' __init__>
   if isinstance(result, sm.RegressionResultsWrapper)
       result = RegressionResults(result)  # wrap with stargazer abstraction
    elif not isinstance(result, StargazerResult):
        raise TypeError("Only subclasses of StargazerResult are allowed")
```

The _minimal_ change needed in stargazer would be to provide the ABC so that it would be as happy to accept a subclasses of the ABC (likely in addition to ResultsWrapper).

> I'm not following you here. This is different from your proposal of abstract class above, right?

stargazer.adapters.abstract would supply StargazerResults, the ABC.  stargazer.adapters.linearmodels would then

```
from stargazer.adapters.abstract implement StargazerResults

class IVResults(StargazerResults):
    def __init__(self, result: "linearmodels.iv.results.IVResults"):
        self._result = result
    
    @property
    def params(self):
        return self._result.params  # pass through, no change

    @property
    def bse(self):
        return self._results.std_errors # different attribute
...
```

`stargazer.adapters.linearmodels.IVResults` is a concrete implementation of the abstract base class `StargazerResults`.

**Comment by toobaz** [comment]
> The most important would be to remove the explicit dependency on RegressionResultsWrapper.

That's for granted, we're discussing how.

> stargazer.adapters.abstract would supply StargazerResults, the ABC. 

OK, I got this

> stargazer.adapters.linearmodels would then

I don't get this. In your view stargazer should provide a base class _and_ the adapters for each library? (this is a no-go) Or should each library be populating the stargazer namespace?

> ... then each library could just expose a same method, e.g. `_modref_get(feature_name)` that stargazer (or anything else) would use.

@bashtage comments on _this_, thanks. It's much simpler than ABC. If you're ready to support it in linearmodels, I'm ready to produce a draft (docs, not code) & support it in stagazer code.

**Comment by bashtage** [comment]
> I don't get this. In your view stargazer should provide a base class _and_ the adapters for each library? (this is a no-go) Or should each library be populating the stargazer namespace?

Where the adapters resid is ultimately up to the stargazer maintainer.  stargazer might find it worth the trouble to provide adpaters for some models, and not for others. For example, it might make sense to provide 1st party support statsmodels GLM results.  

> @bashtage comments on _this_, thanks. It's much simpler than ABC. If you're ready to support it in linearmodels, I'm ready to produce a draft (docs, not code) & support it in stagazer code.

It is not simpler, it is different.  It still needs an ABC so that stargazer would know that it is getting a representation of a more result that it can work with.  I also think it is a nonstarter, since it requires statistical model library writers (and I can speak both as the owner of linearmodels and a committer to statsmodels) to add a "magic" function to support stargazer, and so to take a dependency on stargazer.  I don't think you will get statsmodels, linearmodel, pysal or scikit-learn to do this.  The only libraries that I know that are universal enough to get custom functions to have wide-spread support are NumPy (`__array_function__`) and IPython (`_repr_html_`).

The other approach, which only requires a 3rd party (to each of the statistical libraries) to write an adapter, is much more likely to succeed. This 3rd party library could be stargazer, anything else the community wants to write like `RModels`, or some combination of the two.

**Comment by bashtage** [comment]
> If you're ready to support it in linearmodels, I'm ready to produce a draft (docs, not code) & support it in stagazer code.

Just to 100% clarity, I won't take a dependency on stargazer.  Any adapter that would allow linearmodels results to be used with stargazer would need to live in a different package. I suspect that writing an adapter would be a pretty trivial exercise in most cases.

The upside of writing adapters is that model results classes are extremely stable and almost never get changes in attribute  or property names.

**Comment by toobaz** [comment]
> It still needs an ABC so that stargazer would know that it is getting a representation of a more result that it can work with

``hasattr("_modref_get")`` suffices.

> I won't take a dependency on stargazer

Sorry, I had so far interpreted "dependency" in a literal sense, apparently you mean "any explicit effort towards compatibility with"?

> The other approach, which only requires a 3rd party (to each of the statistical libraries) to write an adapter, is much more likely to succeed.

OK, I will start with ``_modref_get(feature_name)`` but any third party library can easily produce wrappers, including with ABC wrappers if they like.

>  I don't think you will get statsmodels

statsmodels doesn't need an interface as of now, but still, I might write it. I do plan to add one to ``Rmodel`` which, as you correctly observed, is a mess.

**Which features**

Having said the above about _how_ to support libraries, I'd like your help understanding _what_ to support.

The most immediate thing that ``linearmodels`` seems to require on top of ``statsmodels`` is the possible multiplicity of dependent variables. Are there any other obvious ways in which the ``statsmodels`` model is too restrictive, that I should be aware of before trying to design a common interface?

**Comment by bashtage** [comment]
> `hasattr("_modref_get")` suffices.

It doesn't because you don't know what `res._modref_get()` returns. Or at best you have an error-prone  set of possible output.

> Sorry, I had so far interpreted "dependency" in a literal sense, apparently you mean "any explicit effort towards compatibility with"?

No, dependency means that linear models would need to have a test that `_modref_get` produces the correct output in the sense that what it return can be consumed by libraries that it is intended for.  The requires taking an optional dependency (or producing code that isn't tested, which is a no-no for me).

> statsmodels doesn't need an interface as of now, but still, I might write it.

I'm do not think statsmodels will accept it since it creates maintenance burden for them against what may be a fast moving target.  I don't think it belongs _in_ statsmodels.  It is far simpler to have the transformation of the statsmodels results in a 3rd party library (stargazer or anything else). Essentially the chain looks like

```
numpy->scipy & pandas -> statsmodels (and other modeling libraries) -> presentation libraries
```

It shouldn't be 
```
             presentation library -----
                                      |
                                      v
numpy->scipy & pandas -> statsmodels (and other modeling libraries)
```

> Having said the above about how to support libraries, I'd like your help understanding what to support.

It really depends on what end-users what.  Providing a common interface for different statsmodels results types (regression, GLM, Logit/probit, possible some of the time series) would be a great start.  Models from R would also be a natural fit given the fairly uniform structure of the results classes in R. There seems to be interest in linearmodels.  I don't know if there are any PySAL users who would be interested. Ditto for scikit-learn models.

>  Are there any other obvious ways in which the statsmodels model is too restrictive, that I should be aware of before trying to design a common interface?

Some of the statsmodels attribute names are daft. It sometimes uses functions that should be properties. Getting access to some common and important attributes requires 3 or 4 layers.

In terms of a common interface that is designed for a presentation library, what matters is what the presentation library would want to possibly present.  parameters, std errors, t stats, values, R2, Adj R2, f-stat. Probably something similar to what stargazer is pulling from RegressionResultsWrapper only independent of RRW.

**Comment by toobaz** [comment]
> I'm do not think statsmodels will accept it since it creates maintenance burden for them against what may be a fast moving target.

By now, you should have understood that I see it exactly the opposite way: the multiplicity of libraries around, not necessarily very stable, are fast moving targets, while what ``stargazer`` needs and does is pretty obvious. Anyway, ``statsmodels`` is not the problem: ``stargazer`` was born with at least minimal support for it, it is the standard library for statistical models, and stepping back is not an option.

> No, dependency means that linear models would need to have a test that `_modref_get` produces the correct output in the sense that what it return can be consumed by libraries that it is intended for.

You do realize that according to your argument all software standards in the world imply either 1) hard dependencies or 2) untested code... and that it isn't so. You'll have specs to test against, promised.

> It really depends on what end-users what.

I'm really asking you what ``linearmodels`` end-users might want (which is not covered by statsmodels currently). But if there is nothing obvious apart from multiple dependent variables I'll start with that.

**Comment by bashtage** [comment]
> You do realize that according to your argument all software standards in the world imply either 1) hard dependencies or 2) untested code... and that it isn't so. You'll have specs to test against, promised.

There is also bad software, which is what you are describing.  

Good software even with a spec is tested against the spec.  They have to write their own version of the spec which requires mocking when there isn't an abstract implementation available.

In the python universe, an ABC is the programmatic way to convey a spec. See https://docs.python.org/3/library/collections.abc.html which contains a high list of ABCs. Inheriting from one of these ABC is a promise that your subclass will behave in certain ways, so that anyone who consumes a `MutableMapping` know that it is safe to perform certain operations.

> I'm really asking you what `linearmodels` end-users might want (which is not covered by `stargazer` currently). But if there is nothing obvious apart from multiple dependent variables I'll start with that.

I can't answer for them.  This issue seems to want to show an IV2SLS or similar model result in stargazer.

**Comment by toobaz** [comment]
See https://github.com/StatsReporting/stargazer/pull/98
